raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -565,20 +583,62 @@ private FetchQuorumRecordsResponseData buildEmptyFetchQuorumRecordsResponse(
      * - {@link Errors#INVALID_REQUEST} if the request epoch is larger than the leader's current epoch
      *     or if either the fetch offset or the last fetched epoch is invalid
      */
-    private FetchQuorumRecordsResponseData handleFetchQuorumRecordsRequest(
+    private CompletableFuture<FetchQuorumRecordsResponseData> handleFetchQuorumRecordsRequest(
         RaftRequest.Inbound requestMetadata
     ) throws IOException {
         FetchQuorumRecordsRequestData request = (FetchQuorumRecordsRequestData) requestMetadata.data;
+        FetchQuorumRecordsResponseData response = tryCompleteFetchQuorumRecordsRequest(request);
+        if (response.errorCode() != Errors.NONE.code()
+            || response.records().hasRemaining()
+            || request.maxWaitTimeMs() == 0) {
+            return completedFuture(response);
+        }
+
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        purgatory.await(future, request.maxWaitTimeMs());
+
+        return future.handle((nil, exception) -> {
+            if (exception != null) {
+                Throwable cause = exception instanceof ExecutionException ?
+                    exception.getCause() : exception;
+
+                // If the fetch timed out in purgatory, it means no new data is available,
+                // and we will complete the fetch successfully. Otherwise, if there was
+                // any other error, we need to return it.
+                Errors error = Errors.forException(cause);
+                if (error != Errors.REQUEST_TIMED_OUT) {
+                    return buildEmptyFetchQuorumRecordsResponse(error, OptionalLong.empty());
+                }
+            }
+
+            try {
+                return tryCompleteFetchQuorumRecordsRequest(request);
+            } catch (Exception e) {
+                logger.error("Caught unexpected error in fetch completion of request {}", request, e);
+                return buildEmptyFetchQuorumRecordsResponse(Errors.UNKNOWN_SERVER_ERROR, OptionalLong.empty());
+            }
+        });
+    }
+
+    private FetchQuorumRecordsResponseData tryCompleteFetchQuorumRecordsRequest(
```

Comment:
nit: the `validateLeaderOnlyRequest`'s condition is static and hence it seems not necessary to call it twice long with `tryCompleteFetchQuorumRecordsRequest` since we we always trigger `tryCompleteFetchQuorumRecordsRequest` once before putting it into purgatory, maybe we can extract line 626-632 into the `handleFetchQuorumRecordsRequest` to be triggered just once.

Reply:
One of the conditions for completion is becoming a follower. Hence we always have to validate leader status.